Remove RequestDelegateFactory call from RouteEndpointBuilder

halter73 opened the pull request:
- Remove ModelEndpointDataSource. Use RouteEndpointDataSource instead
- Filters will no longer run for razor pages due to this change.

This is somewhat dependent on #42557 to ensure we have a nice experience where "endpoint" filters run on both route handlers, normal RequestDelegates and MVC actions.

Fixes #42703

halter73 committed:
Remove RequestDelegateFactory call from RouteEndpointBuilder

- Remove ModelEndpointDataSource. Use RouteEndpointDataSource instead

davidfowl commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -121,20 +153,25 @@ internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+            // Add MethodInfo and HttpMethodMetadata (if any) as first metadata items as they are intrinsic to the route much like
+            // the pattern or default display name. This gives visibility to conventions like WithOpenApi() to intrinsic route details
+            // (namely the MethodInfo) even when applied early as group conventions.
+            builder.Metadata.Add(handler.Method);
```
I think we want to add this in either case so open api works.

captainsafia replied:
I think adding it only for route handlers will actually help with our goal of making `WithOpenApi` something you can only apply on `RouteHandlerBuilder` types even though we removed that constraint. AFAIK, non-route handler scenarios don't need access to the method info.

halter73 replied:
That's the question. Do we want OpenAPI for these endpoints or not? What if someone manually adds `[Consumes]` or `[Produces]` metadata to these endpoints? For now, I'm trying to keep this behavior consistent with the behavior we've already have.

davidfowl replied:
I think we should unify these endpoints mostly because of the amount of bugs where people end up calling the RequestDelegate overload and it things stop working.

davidfowl commented on src/Http/Routing/test/UnitTests/RouteEndpointBuilderTest.cs:
```diff
@@ -28,4 +28,42 @@ public void Build_AllValuesSet_EndpointCreated()
+    [Fact]
+    public async void Build_DoesNot_RunFilters()
+    {
+
```
```suggestion
```

davidfowl commented on src/Http/Routing/test/UnitTests/Builder/RequestDelegateEndpointRouteBuilderExtensionsTest.cs:
```diff
@@ -233,9 +227,11 @@ public void MapEndpoint_PrecedenceOfMetadata_BuilderMetadataReturned()
         Assert.Equal(3, endpoint.Metadata.Count);
-        Assert.Equal("ATTRIBUTE", GetMethod(endpoint.Metadata[0]));
-        Assert.Equal("METHOD", GetMethod(endpoint.Metadata[1]));
+        Assert.Equal("METHOD", GetMethod(endpoint.Metadata[0]));
+        Assert.Equal("ATTRIBUTE", GetMethod(endpoint.Metadata[1]));
```
Breaking change?

halter73 replied:
Technically.

You'd had to have defined a custom attribute that implemented `IHttpMethodMetadata` which [no one on grepp.app has](https://grep.app/search?q=IHttpMethodMetadata) other than us in this very test, [in one other test](https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/test/testassets/RoutingSandbox/UseEndpointRoutingStartup.cs#L80) which notably didn't need to be changed because it wasn't overriding anything, and in .NET 6 preview diffs before I removed it from the MVC attributes. We certainly don't ship any such attributes.

Plus, we already made this "breaking" change for route handlers between .NET 6 and .NET 7 with #40926. See https://github.com/dotnet/aspnetcore/pull/40926/files#diff-f6727c82b3f6cff19c30c9c793f3e47c1c2ff2035a6453198ff4cedad8e84557

I almost went the other way and had `RouteEndopintDataSource` copy the old .NET 6 behavior here so we didn't "break" anything for either `Delegate` or `RequestDelegate`, but it makes sense to configure metadata added implicitly by the `MapXxx` call like the `MethodInfo` and `HttpMethodMetadata` before any additional metadata configured by the user.

Kahbazi commented on src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:
```diff
@@ -429,18 +404,38 @@ public static RouteHandlerBuilder MapFallback(this IEndpointRouteBuilder endpoin
 
-            dataSource = new RouteEndpointDataSource(endpoints.ServiceProvider, throwOnBadRequest);
-            endpoints.DataSources.Add(dataSource);
+            routeEndpointDataSource = new RouteEndpointDataSource(endpoints.ServiceProvider ?? new EmptyServiceProvider(), throwOnBadRequest);
```
❓ 
```suggestion
            routeEndpointDataSource = new RouteEndpointDataSource(endpoints.ServiceProvider ?? EmptyServiceProvider.Instance, throwOnBadRequest);
```

javiercn commented:
> Filters will no longer run for razor pages due to this change.

This might be problematic as it will break people's expectations that these filters run in both cases like other MVC filters. Is there a specific reason to prevent these filters from applying to pages?

captainsafia commented on src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:
```diff
@@ -186,41 +181,21 @@ public static RouteGroupBuilder MapGroup(this IEndpointRouteBuilder endpoints, R
-        };
-
-        // Add delegate attributes as metadata
-        var attributes = requestDelegate.Method.GetCustomAttributes();
```
I may have missed it but where are we populating metadata from attributes now?

halter73 replied:
`RouteEndpointDataSource` now takes care of that bit since this removed the simpler `ModelEndpointDataSource` which would allow group conventions to unintentionally override endpoint-specific metadata like attributes.

Of course, when conventions are aware of groups, they might want to actually see this metadata as you pointed out #42750. The silver lining here is that these endpoints were never eligible for `.WithOpenApi()` treatment to begin with due to the lack of `MethodInfo` in the metadata, and this much remains unchanged so we don't make that issue any worse. And we're going to fix that issue hopefully anyway.

captainsafia commented on src/Http/Routing/src/RouteEndpointDataSource.cs:
```diff
@@ -233,10 +270,21 @@ private struct RouteEntry
     }
 
+    [Flags]
+    private enum RouteAttributes
```
Nice! I like this.

halter73 committed:
Apply suggestions from code review

Co-authored-by: Kahbazi <A.Kahbazi@gmail.com>
Co-authored-by: David Fowler <davidfowl@gmail.com>

halter73 commented:
> > Filters will no longer run for razor pages due to this change.
> 
> This might be problematic as it will break people's expectations that these filters run in both cases like other MVC filters. Is there a specific reason to prevent these filters from applying to pages?

This is my position as well. The problem is the way they were running in earlier previews is over the `RequestDelegate` (i.e. `Func<HttpContext, Task>` for the `Endpoint`s create for razor pages, and did not provide full fidelity `MethodInfo`s or invocation contexts the way other MVC filters do.

The same is true right now for MVC actions (until after this change merged after which those too will be skipped). @davidfowl opened a draft PR (#42557) to run these filters the way our customers expect for MVC actions. I'm going to take this over since he's on vacation.

We wanted to include razor pages in this PR as well, but the problem is that razor pages can effectively have multiple `MethodInfo`s per endpoint, and that doesn't fit well with our new filter abstraction. We hope to resolve this issue and run filters razor pages in a future release, but we don't know that we can do that in time for .NET 7. We've been considering whether we can "light up" filters for razor pages in future releases without breaking apps.

halter73 committed:
Renames

- Also add MethodInfo to all endpoint metadata in RouteEndpointDataSource

halter73 commented:
I updated this PR to do the renames suggested in #43000 and #42592 which I assigned both to myself.

I also reacted to @davidfowl's and @captainsafia's feedback about adding the `MethodInfo` to the endpoint metadata for old-fashioned `RequestDelegate`-defined endpoints. This will make it show up both with in the `.WithOpenApi()` output and the ApiExplorer stuff we shipped in .NET 6 used by things like swashbuckle and nswag.

Are we sure want this to affect ApiExplorer too? I like that this keeps things consistent with `.WithOpenApi()` but we'll certainly want to make a breaking change announcement.

halter73 commented:
> Are we sure want this to affect ApiExplorer too?

Even if we do agree this is the direction we should take, I think we need an easy way to opt out of this globally for people who might not want this behavior but want to easily upgrade their existing apps using something like swashbuckle or nswag to .NET 7.

Applications might call into libraries that call `MapXXX` internally making adding metadata to these endpoints to the description might be tricky. Doing something like changing `app.MapLibraryEndpoints()` to `app.MapGroup("").IgnoreApi().MapLibraryEndpoints()` if an `IgnoreApi()` extension method existed. If .MapLibraryEndpoints() already returns an `IEndpointConventionBuilder`, you wouldn't even need the group with an empty prefix.

We might want to reopen https://github.com/dotnet/aspnetcore/issues/34068 and add an `IgnoreApi()` extension method that works for both ApiExplorer and WithOpenApi(). @martincostello

halter73 committed:
Fix tests

davidfowl commented on src/Http/Routing/src/RouteEndpointBuilder.cs:
```diff
@@ -12,15 +12,17 @@ namespace Microsoft.AspNetCore.Routing;
+    /// <summary>
+    /// Gets the collection of route handler filter factories associated with this endpoint.
+    /// </summary>
+    public IList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>> RouteHandlerFilterFactories { get; }
```
See https://github.com/dotnet/aspnetcore/pull/42557, this is being moved to the base and the name should be changed. This name is redundant.

halter73 replied:
I did see that. I figured that was just you prototyping in that PR though. Where was the discussion for this? Is there an API proposal? This wasn't mentioned in #42592 in the weeks following its approval.

I agree that removing `RouteHandler` from the name and just calling it `FilterFactories` is better. I do not think this should be moved to base type unless we also add the `RoutePattern` to the base type as well which I don't see happening because of the shadowing. See the following:

https://github.com/dotnet/aspnetcore/blob/8fe74e4d3ef2d26f58e3a2dfef90c64233941cfe/src/Http/Http.Extensions/src/RequestDelegateFactoryOptions.cs#L55-L57

I know you think this creates too tight of a coupling between routing and RDF, but itis very limiting for filters to not have access to the route pattern. `EndpointBuilders` without a `RoutePattern` [don't really exist](https://grep.app/search?q=%3A%20EndpointBuilder&filter[lang][0]=C%23). They're all `RouteEndpointBuilders` because who has time to reimplement route matching without `RoutePatterns`?

Maybe we should unseal `RouteEndpointBuilder`. It seems that this would help [the one non-test, non-inert] `EndpointBuilder`, [ServiceComposer's CompositionEndpointBuilder](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/blob/9eba39a0ab7c7a94751ba6b3e065d5558bf57510/src/ServiceComposer.AspNetCore/CompositionEndpointBuilder.cs) which really want to be a `RouteEndpointBuilder` considering it has a `RoutePattern` it cannot expose and it returns a `RouteEndpoint` from `Build()`. @mauroservienti

halter73 replied:
When I made that comment in RequestDelegateFactoryOptions and when leaving my last comment minutes ago, I forgot that the direction of the Routing -> Http.Extension dependency makes it impossible to expose `RoutePattern`s to filters without some crazy type forwarding. That is sad. Maybe we'll just expose it via metadata 😆

I still think we should just unseal `RouteEndpointBuilder` and leave `FilterFactories` there.

davidfowl replied:
I don't see why we would leave it on the derived type when the type got renamed to Endpoint* and not Route*. It has nothing to do with routing at this point.

halter73 replied:
Now that I realize that the RoutePattern cannot be exposed to the filters because of the layering, I feel less strongly about it being in RouteEndopintBuilder. I'll update the API proposals to match what's in https://github.com/dotnet/aspnetcore/pull/42557 before our normal Monday API review.

I wish the routing stuff was lower level though. I still haven't seen a use case for EndpointBuilders or Endpoints that are not a RouteEndopintBuilders or RouteEndpoints. Having to downcast all the time is annoying. Are there any examples of this? Or is it more of a just-in-case thing? I realize you could in theory do your own matching based on something other than the RoutePattern. I just haven't seen it.

captainsafia replied:
The general use of `EndpointBuilder` as a context object makes sense to me. Is this use of builders-as-context-objects common anywhere else? The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

halter73 replied:
> Is this use of builders-as-context-objects common anywhere else?

I'd consider it more of a builders-in-context-objects, but I don't know of another example in ASP.NET. Contexts and builders as separate parameters are more common, but that's for callbacks that are specifically intended to add to the builder.

It might actually make sense to change `IEndpointMetadataProvider.PopulateMetadata(EndpointMetadataContext)` to `IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder)` and `IEndpointParameterMetadataProvider.PopulateMetadata(EndpointParameterMetadataContext)` to `IEndpointParameterMetadataProvider.PopulateMetadata(ParameterInfo, EndpointBuilder)` and get rid of the new contexts entirely.

For filters, we could change the factory type from `Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>` to `Func<MethodInfo, EndopintBuilder, EndpointFilterDelegate, EndpointFilterDelegate>` and get rid of the `EndpointFilterFactoryContext`.

In the filter case, I like the flexibility of the context though, and the primary purpose of filters is probably not going to be to update metadata on the `EndpointBuilder`. This is the whole point of the meta data providers which is why I support removing the contexts there. @DamianEdwards 

> The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

I don't think this will be confusing because it's not like we're exposing an `IEndpointConventionBuilder` which has a bunch of extension methods targeting it that add conventions. The `EndpointBuilder` does make it easy to modify metadata which many conventions do, but filters and metadata providers are still going to have to add the metadata manually rather than use extension methods just like today.

halter73 commented on src/Shared/RoslynUtils/TypeHelper.cs:
```diff
@@ -36,4 +37,33 @@ internal static bool IsCompilerGeneratedMethod(MethodInfo method)
+    /// <summary>
+    /// Tries to get non-compiler-generaged name of function. This parses generated local function names out of a generated method name if possible.
+    /// </summary>
+    internal static bool TryGetNonCompilerGeneratedMethodName(MethodInfo method, [NotNullWhen(true)] out string? originalName)
```
This is shared source, but I think this is the only caller so I can make it private. I like calling the method over doing `Attribute.IsDefined(...) || IsCompilerGeneratedType(...)` inline, because I think it's a little more self-documenting.

captainsafia commented on src/Http/Routing/src/ConfigureRouteHandlerOptions.cs:
```diff
@@ -8,16 +8,16 @@ namespace Microsoft.AspNetCore.Routing;
+    private readonly IHostEnvironment? _environment;
 
-    public ConfigureRouteHandlerOptions(IHostEnvironment environment)
+    public ConfigureRouteHandlerOptions(IHostEnvironment? environment = null)
```
What's the motivation for this change?

halter73 replied:
It's to avoid breaking unit tests that add options to the service collection but not IHostEnvironment. There were a few tests in downstream projects like SignalR that were broken by this. This was a simple fix that might also avoid breaking external tests and odd configurations without an IHostEnvironment.

halter73 committed:
Fix tests and address PR feedback

- No. I do not think normalizing the display names is a breaking change.

halter73 commented:
I think I've addressed all the PR feedback.

I undid the renames to reduce merge conflicts in #42557. The renames weren't the point of this PR. I thought I was saving work, but I wasn't. We also want to expose the `EndopintBuilder` in more places (see https://github.com/dotnet/aspnetcore/issues/43000#issuecomment-1201920724). This will be done in another PR.

halter73 committed:
Revert "Renames"

This reverts commit bb2b157c3b6db7c33543b6f8622ce9a5ef08f8e8.

halter73 committed:
Add MethodInfo to all endpoint metadata in RouteEndpointDataSource

halter73 committed:
Add back TryParseLocalFunctionName

halter73 committed:
internal -> private

halter73 committed:
invocationCallCount is 1 in Build_DoesNot_RunFilters test
